internal/util/reftracker/reftracker.go

```diff
@@ -0,0 +1,162 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reftracker
+
+import (
+	goerrors "errors"
+	"fmt"
+
+	"github.com/ceph/go-ceph/rados"
+
+	"github.com/ceph/ceph-csi/internal/util/reftracker/errors"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/radoswrapper"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/reftype"
+	v1 "github.com/ceph/ceph-csi/internal/util/reftracker/v1"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/version"
+)
+
+// reftracker is key-based implementation of a reference counter.
+//
+// Unlike integer-based counter, reftracker counts references by tracking
+// unique keys. This allows accounting in situations where idempotency must be
+// preserved. It guarantees there will be no duplicit increments or decrements
+// of the counter.
+//
+// It is stored persistently as a RADOS object, and is safe to use with
+// multiple concurrent writers, and across different nodes of a cluster.
+
+// Add atomically adds refs to reference tracker named `rtName`.
+// If the reftracker object doesn't exist yet, it is created. This is also
+// reflected in the return value.
+func Add(
+	ioctx radoswrapper.IOContextW,
+	rtName string,
+	refs map[string]struct{},
+) (created bool, err error) {
+	if err = validateAddInput(rtName, refs); err != nil {
+		return false, err
+	}
+
+	// Read reftracker version.
+
+	rtVer, err := version.Read(ioctx, rtName)
+	if err != nil {
+		if goerrors.Is(err, rados.ErrNotFound) {
+			// This is a new reftracker. Initialize it with `refs`.
+			if err = v1.Init(ioctx, rtName, refs); err != nil {
+				return false, fmt.Errorf("failed to initialize reftracker: %w", err)
+			}
+
+			return true, nil
+		}
+
+		return false, fmt.Errorf("failed to read reftracker version: %w", err)
+	}
+
+	// Add references to reftracker object.
+
+	gen, err := ioctx.GetLastVersion()
+	if err != nil {
+		return false, fmt.Errorf("failed to get RADOS object version: %w", err)
+	}
+
+	switch rtVer {
+	case v1.Version:
+		err = v1.Add(ioctx, rtName, gen, refs)
+		if err != nil {
+			err = fmt.Errorf("failed to add refs: %w", err)
+		}
+	default:
+		err = errors.UnknownObjectVersion(rtVer)
+	}
+
+	return false, err
+}
+
+// Remove atomically removes refs from reference tracker named `rtName`.
+// If the reftracker object holds no references after this removal, the whole
+// object is deleted too. This is also reflected in the return value.
+// If the reftracker object doesn't exist, (true, nil) is returned.
+func Remove(
+	ioctx radoswrapper.IOContextW,
+	rtName string,
+	refs map[string]reftype.RefType,
+) (deleted bool, err error) {
```

Comment:
no need for named returns? same in other places also

Reply:
I think it helps with readability to be explicit about what this boolean represents. It returns `true` only if the whole object was deleted, not if a reference in the object was removed. I think it's an important distinction, and having it embedded in the function signature makes this a bit more obvious I hope.

The comment of the function also mentions removal and deletion as two separate things:
> If the reftracker object holds no references after this **removal**, the whole
> object is **deleted** too.

I'll leave it for now, but please let me know if you insist on removing it anyway.

Reply:
We dont use named returns anymore (we stopped long back). mixing named/naked returns is not a good idea for the whole repo.